src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,240 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency"""
+
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar
+
+import numpy as np
+
+T = TypeVar("T")
+
+def RingBufferArray(size: int):
+    return RingBuffer[T](np.empty(shape=(size,), dtype=np.float64))
+
+
+
+class RingBuffer(Generic[T]):
+    def __init__(self, container: Any):
+        """Initialize the ring buffer with the given container.
+
+        Args:
+            container: Container to store the data in.
+        """
+        self.container = container
+        self.write_index = 0
+        self.read_index = 0
+        self.size = 0
+
+    def __len__(self):
+        """Return the amount of items that this container currently holds."""
+        return self.size
+
+    def max_size(self):
+        """Return the max amount of items this container can hold"""
+        return len(self.container)
+
+    def push(self, value: T) -> int:
+        """Pushes a new value into the ring buffer, returns the index in the
+        ringbuffer.
+
+        Args:
+            value: value to push into the ring buffer.
+        """
+
+        if self.size == len(self.container):
+            # Move read position one forward, dropping the oldest written value
+            self.read_index = (self.read_index + 1) % len(self.container)
+        else:
+            self.size += 1
+
+        self.container[self.write_index] = value
+        value_index = self.write_index
+        self.write_index = (self.write_index + 1) % len(self.container)
+
+        return value_index
+
+    def pop(self) -> T:
+        """Remove the oldest value from the ring buffer and return it."""
+
+        if self.size == 0:
+            raise
+
+        val = self.container[self.read_index]
+        self.read_index = (self.read_index + 1) % len(self.container)
+        self.size -= 1
+
+        return val
+
+    def wrap(self, index) -> int:
+        return index % len(self.container)
+
+    def full(self) -> bool:
+        return len(self) == len(self.container)
+
+    def __setitem__(self, index: int | slice, value: T | Sequence[T]) -> None:
+        self.container[index] = value
+
+    def __getitem__(self, index_or_slice: int | slice) -> T | Sequence[T]:
+        """Returns the value at the given index or value range at the given
+        slice. Does not support wrap-around or copying of data."""
+        return self.container[index_or_slice]
+
+
+class OrderedRingBuffer(Generic[T]):
+    """Time aware ringbuffer that keeps its entries sorted time."""
```

Comment:
Hu? You mean the whole class is not necessary? It is clearly specified in the ticket that a sorted ringbuffer is requested...

Reply:
The use case I documented there was for resampling, in case at some point we get out-of-order data from a data source, but it is a hypothetical very low priority case. Sorry if it is not clear in the issue, I think we mentioned it in live conversations but we should have made it more clear in the issue too if it's not.

In any case it looks like @matthias-wende-frequenz might have a use case for this in the prediction pipeline.